tests/clone/clone.cpp

```diff
@@ -0,0 +1,550 @@
+
+    auto cClone = c->clone();
+
+    compareComponent(c, cClone);
+}
+
+TEST(Clone, componentWithChildren)
+{
+    auto c = libcellml::Component::create();
+    auto c1 = libcellml::Component::create();
+    auto c2 = libcellml::Component::create();
+
+    c->setId("unique");
+    c->setName("copy");
+
+    c->addComponent(c1);
+    c->addComponent(c2);
+
+    auto cClone = c->clone();
+
+    compareComponent(c, cClone);
+}
+
+TEST(Clone, componentWithImport)
+{
+    auto c = libcellml::Component::create();
+    auto import = libcellml::ImportSource::create();
+    import->setUrl("some-other-model.xml");
+
+    c->setId("unique");
+    c->setName("copy");
+    c->setSourceComponent(import, "imported_component_name");
+
+    auto cClone = c->clone();
+
+    compareComponent(c, cClone);
+}
+
+TEST(Clone, componentWithImportAndChildren)
+{
+    auto c = libcellml::Component::create();
+    auto c1 = libcellml::Component::create();
+    auto c2 = libcellml::Component::create();
+    auto import = libcellml::ImportSource::create();
+
+    import->setUrl("some-other-model.xml");
+    c->setId("unique");
+    c->setName("copy");
+    c1->setName("child_1");
+    c2->setName("child_2");
+
+    c->addComponent(c1);
+    c->addComponent(c2);
+    c->setSourceComponent(import, "my_component_name");
+
+    auto cClone = c->clone();
+
+    compareComponent(c, cClone);
+}
+
+void compareModel(const libcellml::ModelPtr &m1, const libcellml::ModelPtr &m2)
+{
+    EXPECT_EQ(m1->id(), m2->id());
+    EXPECT_EQ(m1->name(), m2->name());
+
+    EXPECT_EQ(m1->unitsCount(), m2->unitsCount());
+    EXPECT_EQ(m1->componentCount(), m2->componentCount());
+
+    for (size_t index = 0; index < m1->unitsCount(); ++index) {
+        auto u1 = m1->units(index);
+        auto u2 = m2->units(index);
+        compareUnits(u1, u2, m2);
+    }
+
+    for (size_t index = 0; index < m1->componentCount(); ++index) {
+        auto c1 = m1->component(index);
+        auto c2 = m2->component(index);
+        compareComponent(c1, c2, m2);
+    }
+}
+
+TEST(Clone, model)
+{
+    auto m = libcellml::Model::create();
+    m->setId("unique_model");
+    m->setName("model");
+    m->setEncapsulationId("encapsulation_id");
+
+    auto mClone = m->clone();
+
+    compareModel(m, mClone);
+}
+
+TEST(Clone, modelWithUnits)
+{
+    auto m = libcellml::Model::create();
+    m->setId("unique_model");
+    m->setName("model");
+
+    auto mClone = m->clone();
+
+    compareModel(m, mClone);
+}
+
+TEST(Clone, modelWithComponents)
+{
+    auto m = libcellml::Model::create();
+    m->setId("unique_model");
+    m->setName("model");
+
+    auto c = libcellml::Component::create();
+    auto c1 = libcellml::Component::create();
+    auto c2 = libcellml::Component::create();
+
+    c->setId("unique");
+    c->setName("copy");
+    c1->setName("child_1");
+    c2->setName("child_2");
+
+    c->addComponent(c1);
+    c->addComponent(c2);
+
+    auto mClone = m->clone();
+
+    compareModel(m, mClone);
+}
```

Comment:
Shouldn't `c` be added as a component to `m`?

Reply:
Double O for awesome, I'm on fire.

Reply:
Makes me wonder whether I haven't missed other similar awesomeness from you... :)

Reply:
That has to be the limit of my awesomeness! surely 😆

Reply:
I would hope! Hmm... :)